[spirv] Add support for function calls.

denis0x0D opened the pull request:
This patch is addressing https://github.com/tensorflow/mlir/issues/131

@antiagainst @MaheshRavishankar can you please take a look?
Thanks!

Commit:
[spirv] Add support for function calls.

Add spv.FunctionCall operation and (de)serialization.

antiagainst commented on lib/Dialect/SPIRV/Serialization/Serializer.cpp:
```diff
@@ -517,8 +547,80 @@ Serializer::processMemberDecoration(uint32_t structID, uint32_t memberIndex,
 } // namespace
 
+LogicalResult Serializer::processFunctionCallOp(spirv::FunctionCallOp op) {
+  auto funcAttrSymbolRef = op.getAttrOfType<SymbolRefAttr>("callee");
```
`op.callee()` ?

antiagainst commented on lib/Dialect/SPIRV/Serialization/Serializer.cpp:
```diff
@@ -517,8 +547,80 @@ Serializer::processMemberDecoration(uint32_t structID, uint32_t memberIndex,
+
+  for (auto *value : op.arguments()) {
+    auto valueID = findValueID(value);
+    if (!valueID) {
```
The IR is verified so this cannot happen? I think we can turn it into an `assert` here.

antiagainst commented on lib/Dialect/SPIRV/Serialization/Serializer.cpp:
```diff
@@ -517,8 +547,80 @@ Serializer::processMemberDecoration(uint32_t structID, uint32_t memberIndex,
+    valueIDMap[op.getResult(0)] = funcCallID;
+  }
+
+  // The spec allows to invoke FunctionCallOp as a forward call.
```
Instead of doing this, what about having a `getOrCreateFunctionID(StringRef)` function, which internally uses `getNextID` as the ID for that function if it is not seen yet? I think that will be simpler.

denis0x0D replied:
oh, I see, thanks a lot!

antiagainst commented on lib/Dialect/SPIRV/Serialization/Serializer.cpp:
```diff
@@ -1374,6 +1476,9 @@ LogicalResult Serializer::processOperation(Operation *op) {
   if (auto constOp = dyn_cast<spirv::ConstantOp>(op)) {
     return processConstantOp(constOp);
   }
+  if (auto funcCallOp = dyn_cast<spirv::FunctionCallOp>(op)) {
```
Similarly, if we do `processOp<spirv::FunctionCallOp>` we don't need to explicitly register here.

MaheshRavishankar replied:
+1

antiagainst commented on lib/Dialect/SPIRV/Serialization/Serializer.cpp:
```diff
@@ -517,8 +547,80 @@ Serializer::processMemberDecoration(uint32_t structID, uint32_t memberIndex,
+  auto funcName = funcAttrSymbolRef.getValue();
+  auto funcID = findFunctionID(funcName);
+  auto funcCallID = getNextID();
+  SmallVector<uint32_t, 3> operands{resTypeID, funcCallID, funcID};
```
8 instead of 3 to avoid allocation for the common cases?

antiagainst commented on lib/Dialect/SPIRV/Serialization/Serializer.cpp:
```diff
@@ -517,8 +547,80 @@ Serializer::processMemberDecoration(uint32_t structID, uint32_t memberIndex,
+    ++argIndex;
+  }
+
+  if (resultTypes.size()) {
```
!resultTypes.empty()

antiagainst commented on test/Dialect/SPIRV/control-flow-ops.mlir:
```diff
@@ -144,6 +144,119 @@ func @weights_cannot_both_be_zero() -> () {
+    spv.FunctionCall @f_2() : () -> ()
+    // CHECK: {{%.*}} = spv.FunctionCall @f_4({{%.*}}) : (i32) -> i32
+    %1 = spv.FunctionCall @f_4(%arg2) : (i32) -> i32
+    // CHECK: {{%.*}} = spv.FunctionCall @f_4({{%.*}}) : (i32) -> i32
```
What's the purpose of this check? Seems the same as above.

antiagainst commented on test/Dialect/SPIRV/control-flow-ops.mlir:
```diff
@@ -144,6 +144,119 @@ func @weights_cannot_both_be_zero() -> () {
+  }
+
+  func @f_3(%arg0 : !spv.struct<!spv.array<4xf32>>) -> (!spv.struct<!spv.array<4xf32>>) {
+    // CHEKC: {{%.*}} = spv.FunctionCall @f_3({{%.*}}) : (!spv.struct<!spv.array<4xf32>>) -> !spv.struct<!spv.array<4xf32>>
```
Typo: CHECK.

Do we need this case? Seems already covered by fmain. (I mean we don't check new stuff here right?)

antiagainst commented on test/Dialect/SPIRV/Serialization/function_call.mlir:
```diff
@@ -0,0 +1,68 @@
+    spv.FunctionCall @f_2() : () -> ()
+    // CHECK: spv.FunctionCall @f_3({{%.*}}) : (!spv.ptr<!spv.array<4 x f32>, Input>) -> !spv.ptr<!spv.array<4 x f32>, Input>
+    %4 = spv.FunctionCall @f_3(%1) : (!spv.ptr<!spv.array<4xf32>, Input>) -> !spv.ptr<!spv.array<4xf32>, Input>
+    // CHECK: spv.FunctionCall @f_3({{%.*}}) : (!spv.ptr<!spv.array<4 x f32>, Input>) -> !spv.ptr<!spv.array<4 x f32>, Input>
```
Do we need this case? It doesn't cover something new compared to the above one.

denis0x0D replied:
I was thinking to check, that we can pass a result value from FunctionCall as operand for the other FunctionCall.

antiagainst replied:
The result of a function call is a normal SSA value, so nothing different from other SSA values. I don't think we need to check it. :)

denis0x0D replied:
@antiagainst it's just check fro my first invalid implementation, I was missing to put and id of OpFunctionCall into id valueMap for serializer. I'll delete, thanks!

antiagainst replied:
Oh, I see. In that case, you can `spv.Return %3 : i32` or something. No need to create another `spv.FunctionCall` just to use the result of a `spv.FunctionCall`. :)

denis0x0D replied:
@antiagainst thanks, got it!

antiagainst commented on test/Dialect/SPIRV/Serialization/function_call.mlir:
```diff
@@ -0,0 +1,68 @@
+    spv.Return
+  }
+  func @f_2() -> () {
+    // CHECK: spv.FunctionCall @f_2() : () -> ()
```
This one is not needed. It's already covered in fmain.

antiagainst commented on test/Dialect/SPIRV/Serialization/function_call.mlir:
```diff
@@ -0,0 +1,68 @@
+
+  func @f_4() -> () {
+    %count = spv.constant 16 : i32
+    // CHECK: spv.FunctionCall @f_loop_with_function_call({{%.*}}) : (i32) -> ()
```
Don't need this one? It's already covered in fmain.

denis0x0D replied:
I was thinking to check that Function call can work under control flow ops such as loops.

antiagainst replied:
But that's for the next function?

denis0x0D replied:
@antiagainst your are right! Thanks!

antiagainst reviewed (changes requested):
Awesome! Thanks for adding support for this, Denis!

denis0x0D commented on lib/Dialect/SPIRV/SPIRVOps.cpp:
```diff
@@ -912,6 +912,116 @@ static void print(spirv::ExecutionModeOp execModeOp, OpAsmPrinter *printer) {
+}
+
+static void print(spirv::FunctionCallOp functionCallOp, OpAsmPrinter *printer) {
+  SmallVector<Type, 0> argTypes(functionCallOp.getOperandTypes());
```
@antiagainst can you please clarify, it seems like it uses under L962, we have to print function type, please fix me if I'm wrong.

antiagainst replied:
Ah, right. Sorry, my bad.

antiagainst commented on lib/Dialect/SPIRV/Serialization/Deserializer.cpp:
```diff
@@ -500,6 +503,70 @@ LogicalResult Deserializer::processExtension(ArrayRef<uint32_t> operands) {
+      arguments);
+
+  if (resultTypes.size()) {
+    valueMap[resultID] = opFunctionCall.getResult(0);
```
Yes, you are correct. Please ignore this comment. :)

denis0x0D commented on lib/Dialect/SPIRV/SPIRVOps.cpp:
```diff
@@ -912,6 +912,116 @@ static void print(spirv::ExecutionModeOp execModeOp, OpAsmPrinter *printer) {
+           << funcType.getNumResults();
+  }
+
+  if (parser->addTypesToList(funcType.getResults(), state->types) ||
```
@antiagainst can you please clarify, sorry, I probably didn't get it right, parser returns an instance of ParseResult, logical operator || returns bool, I, probably missing the convertion from bool to ParseResult.

River707 replied:
failure(parser->... || parser->...)

antiagainst replied:
Sorry I meant `return failure(parser->... || parser->...)`.

denis0x0D replied:
got it! Thanks!

denis0x0D replied:
@MaheshRavishankar thanks for mention it, I was actually learning from standard ops and llvm ops.

denis0x0D commented on include/mlir/Dialect/SPIRV/SPIRVControlFlowOps.td:
```diff
@@ -151,6 +151,53 @@ def SPV_BranchConditionalOp : SPV_Op<"BranchConditional", [Terminator]> {
 
 // -----
 
+def SPV_FunctionCallOp : SPV_Op<"FunctionCall"> {
```
@antiagainst in this case should InFunctionScope be changed 
from 
`CPred<"llvm::isa_and_nonnull<FuncOp>($_op.getParentOp())">>;`
to something like this
`CPred<"llvm::isa_and_nonnull<FuncOp>($_op.getParentOfType<FuncOp>())">>;`

If we put op with this trait inside the loop op, we will get the LoopOp as parent, please fix me if I'm wrong.

denis0x0D replied:
Updated:
`CPred<"($_op.getParentOfType<FuncOp>())">>;`

MaheshRavishankar reviewed (commented):
Thanks a lot @denis0x0D ! This is really useful

denis0x0D commented on lib/Dialect/SPIRV/SPIRVOps.cpp:
```diff
@@ -912,6 +912,116 @@ static void print(spirv::ExecutionModeOp execModeOp, OpAsmPrinter *printer) {
+  FunctionType type;
+  SmallVector<OpAsmParser::OperandType, 4> operands;
+  auto loc = parser->getNameLoc();
+  if (parser->parseAttribute(calleeAttr, "callee", state->attributes) ||
```
Oh, I see, thanks for mention this!

denis0x0D commented:
@antiagainst @MaheshRavishankar thanks a lot for review! I've updated the patch, can you please take a look?
Thanks.

antiagainst commented on lib/Dialect/SPIRV/SPIRVOps.cpp:
```diff
@@ -912,6 +913,108 @@ static void print(spirv::ExecutionModeOp execModeOp, OpAsmPrinter *printer) {
+}
+
+static void print(spirv::FunctionCallOp functionCallOp, OpAsmPrinter *printer) {
+  SmallVector<Type, 0> argTypes(functionCallOp.getOperandTypes());
```
4 instead of 0?

antiagainst commented on lib/Dialect/SPIRV/SPIRVOps.cpp:
```diff
@@ -912,6 +913,108 @@ static void print(spirv::ExecutionModeOp execModeOp, OpAsmPrinter *printer) {
+
+static void print(spirv::FunctionCallOp functionCallOp, OpAsmPrinter *printer) {
+  SmallVector<Type, 0> argTypes(functionCallOp.getOperandTypes());
+  SmallVector<Type, 0> resultTypes(functionCallOp.getResultTypes());
```
1 instead of 0?

denis0x0D replied:
Oh, sorry, i've missed it again.

antiagainst commented on lib/Dialect/SPIRV/Serialization/Deserializer.cpp:
```diff
@@ -1779,6 +1779,53 @@ Deserializer::processOp<spirv::ExecutionModeOp>(ArrayRef<uint32_t> words) {
+    auto *value = getValue(operand);
+    if (!value) {
+      return emitError(unknownLoc, "unknown <id> ")
+             << operand << " used by spv.FunctionCall";
```
OpFunctionCall instead of spv.FunctionCall. Here we are deserializing SPIR-V blob so the error message should be talking about SPIR-V instructions.

denis0x0D replied:
Thanks, fixed!

antiagainst commented on lib/Dialect/SPIRV/Serialization/Deserializer.cpp:
```diff
@@ -1779,6 +1779,53 @@ Deserializer::processOp<spirv::ExecutionModeOp>(ArrayRef<uint32_t> words) {
+  auto resultID = operands[1];
+  auto functionID = operands[2];
+  
+  auto functionName = nameMap.lookup(functionID).str();
```
Can we have a `getFunctionSymbol` similar to `getSpecConstantSymbol` so the same logic can be reused in `processFunction`?

antiagainst commented on lib/Dialect/SPIRV/Serialization/Serializer.cpp:
```diff
@@ -131,6 +131,15 @@ class Serializer {
     return funcIDMap.lookup(fnName);
   }
 
+  uint32_t getOrCreateFunctionID(StringRef fnName) {
```
We only use inline implementation when it's just a trivial oneliner. Can you move the implementation here out side of the decl?

denis0x0D replied:
Fixed, thanks!

antiagainst commented on lib/Dialect/SPIRV/Serialization/Serializer.cpp:
```diff
@@ -131,6 +131,15 @@ class Serializer {
     return funcIDMap.lookup(fnName);
   }
 
+  uint32_t getOrCreateFunctionID(StringRef fnName) {
```
Can you put a doc for this function? "// Gets the <id> for the function with the given name. Assigns the next available <id> if the function haven't been deserialized."

antiagainst reviewed (changes requested):
Nice, thanks a lot! Just a few more nits. :)

denis0x0D commented:
@antiagainst thanks a lot for review, I've updated the patch, I also want to note, that I've updated the InFunctionScope trait, from

`CPred<"llvm::isa_and_nonnull<FuncOp>($_op.getParentOp())">>;
`

to 

`CPred<"($_op.getParentOfType<FuncOp>())">>;`

because previous one does not work if op is inside control flow op, I hope it's right.
Thanks.

antiagainst commented:
@denis0x0D: yes, that makes sense. All those ops marked as `InFunctionScope` should be allowed in loop regions too. Thanks for fixing this! :)

antiagainst commented:
Another recommendation regarding PR workflow: it would be nice if you can address comments by adding additional commits to the feature branch instead of force pushing. It would make code review easier so we can land PRs faster. :) We can do force pushing at the end after approval. Thanks! :)

denis0x0D commented:
> Another recommendation regarding PR workflow: it would be nice if you can address comments by adding additional commits to the feature branch instead of force pushing. It would make code review easier so we can land PRs faster. :) We can do force pushing at the end after approval. Thanks! :)

thanks a lot for mention this!

walli55 reviewed (commented):
Alomaisi
